text/0121-config-service.md

```diff
@@ -0,0 +1,161 @@
+    // This proposal focuses on the metrics portion, but tracing is a natural
+    // candidate for something that might be useful to configure in the future,
+    // so this field will be left here for that eventuality.
+  }
+  TraceConfig trace_config = 3;
+
+  // Optional. The client is suggested to wait this long (in seconds) before
+  // pinging the configuration service again.
+  int32 suggested_wait_time_sec = 4;
+}
+```
+
+The SDK will periodically read a config from the service using GetConfig. If it fails to do so, it will just use either the default config or the most recent successfully read config. If it reads a new config, it will apply it.
+
+Export frequency from the SDK depends on Schedules. There can only be one Schedule for now, which defines the schedule for all metrics. The schedule has a CollectionPeriod, which defines how often metrics are exported.
+
+In the future, we will add per-metric configuration. Each Schedule also has inclusion_patterns and exclusion_patterns. Any metrics that match any of the inclusion_patterns and do not match any of the exclusion_patterns will be exported every CollectionPeriod (ie. every minute). A component will be added that can export metrics that match a pattern from one Schedule at that Schedule's collection period, while exporting other metrics that match the patterns of another Schedule at that other Schedule's collection period.
+
+The collector will support a new interface for a DynamicConfig service that can be used by an SDK, allowing a custom implementation of the configuration service protocol described above, to act as an optional bridge between an SDK and an arbitrary configuration service. This interface can be implemented as a shim to support accessing remote configurations from arbitrary backends. The collector is configured to expose an endpoint for requests to the DynamicConfig service, and returns results on that endpoint.
+
+## Trade-offs and mitigations
+
+This feature will be implemented purely as an experiment, to demonstrate its viability and usefulness. More investigation can be done after a rough prototype is demonstrated.
+
+As mentioned [here](https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048), the configuration service can be a potential attack vector for an application instrumented with Open Telemetry, depending on what we allow in the protocol. This can be mitigated since this proposal only allows the SDK to read from a configuration service managed by the collector. Additionally, we can highlight in the remote configuration protocol that for future changes, caution is needed in terms of the sorts of configurations we allow.
+
+## Prior art and alternatives
+
+One way to configure metric export schedules is to, instead of pushing the metrics, use a pull mechanism to have metrics pulled whenever wanted. This has been implemented for [Prometheus](https://github.com/open-telemetry/opentelemetry-go/pull/751).
```

Comment:
This is an alternative to the current push-based approach to metrics, but not really an alternative to the proposal here. Echoing @tigrannajaryan's comment above, it sounds like the important alternative to note here is pushing vs. polling the config.

On its face, It's pretty surprising that we'd ask end users to implement a remote configuration service which the collector (or SDK) would poll, instead of adding something like a `setConfig` API method to the collector. Why did you decide on polling here?

Reply:
Our primary rationale was the large-scale use case where the collector would be responsible for pushing configs to many instrumented processes. It seemed more scalable to have the SDK read from the configuration service.

In a situation where the collector is extended to act as a configuration service or act as a proxy for remote configuration built into a monitoring backend, hopefully the end user won't be need to do much extra work relative to setting up a mechanism that calls a `setConfig` method on the collector.

The trade-offs for pushing vs. polling is definitely something worth thinking about though.